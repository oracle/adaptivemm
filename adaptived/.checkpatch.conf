# There is no Kernel tree available, run without it.
--no-tree

# Ignore the SPDX License tag, that's supposed to be the first line in
# Linux Kernel sources.
--ignore SPDX_LICENSE_TAG

# Follow the 100-character limit.
--max-line-length=100

# Kernel specific kstr* functions recommendation doesn't apply to us.
--ignore SSCANF_TO_KSTRTO

# Ignore Kernel str* function recommendations
--ignore STRCPY
--ignore STRNCPY

# Ignore the suggestion to add file changes (add, move, or delete) to the
# MAINTAINERS file
--ignore FILE_PATH_CHANGES
