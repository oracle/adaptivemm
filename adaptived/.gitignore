*~
*.css
*.la
*.lo
*.o
*.pc
*.swp
*.orig
*.gcno
*.gcda
*.html
*.info
*.log
*.trs
*-coverage/
.deps
.dirstamp
.libs
configure
Makefile
Makefile.in
cscope.*
tags
gtest
/autom4te.cache
/aclocal.m4
/build-aux
/config.*
/configure
/configure.h*
/libtool
/m4
/stamp-h1
/cov-int
/src/adaptived
