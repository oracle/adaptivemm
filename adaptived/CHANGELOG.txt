adaptived: Releases
===============================================================================
https://github.com/oracle/adaptivemm

* Version 1.0.1 - February 24, 2025
- Fix bug where only two or fewer rules could be loaded
- Fix duration bug in PSI cause
- Fix memory leak reported by coverity

* Version 1.0.0 - November 8, 2024
- Initial release
