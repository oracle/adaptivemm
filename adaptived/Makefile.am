ACLOCAL_AMFLAGS = -I m4
DIST_SUBDIRS = dist doc include src tests
SUBDIRS = ${DIST_SUBDIRS}

EXTRA_DIST = \
	CONTRIBUTING.md \
	LICENSE.txt \
	README.md \
	SECURITY.md \
	THIRD_PARTY_LICENSES.txt

help:
	@echo "adaptived build system"
	@echo " make targets:"
	@echo "  (none):           build the library"
	@echo "  check:            build the library and run tests"
	@echo "  check-build:      build the library and all tests"

check-build: all
	${MAKE} ${AM_MAKEFLAGS} -C src check-build
	${MAKE} ${AM_MAKEFLAGS} -C tests check-build
