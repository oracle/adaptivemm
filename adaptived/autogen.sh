#!/bin/bash
#

git submodule update --init --recursive

# configure googletest
pushd googletest/googletest
git checkout release-1.8.0
cmake -DBUILD_SHARED_LIBS=ON .
make
popd

test -d m4 || mkdir m4
autoreconf -fi
rm -rf autom4te.cache
