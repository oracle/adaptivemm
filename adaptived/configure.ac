dnl # Copyright (c) 2024, Oracle and/or its affiliates.
dnl # DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
dnl #
dnl # This code is free software; you can redistribute it and/or modify it
dnl # under the terms of the GNU General Public License version 2 only, as
dnl # published by the Free Software Foundation.
dnl #
dnl # This code is distributed in the hope that it will be useful, but WITHOUT
dnl # ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
dnl # FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
dnl # version 2 for more details (a copy is included in the LICENSE file that
dnl # accompanied this code).
dnl #
dnl # You should have received a copy of the GNU General Public License version
dnl # 2 along with this work; if not, write to the Free Software Foundation,
dnl # Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
dnl #
dnl # Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
dnl # or visit www.oracle.com if you need additional information or have any
dnl # questions.

AC_INIT([adaptived],[0.0.0])

dnl #
dnl # autoconf configuration
dnl #
AC_CONFIG_AUX_DIR([build-aux])
AC_CONFIG_HEADERS([configure.h])
AC_CONFIG_MACRO_DIR([m4])
AM_INIT_AUTOMAKE(-Wall foreign subdir-objects)

m4_ifdef([AM_SILENT_RULES], [AM_SILENT_RULES([yes])])
m4_ifdef([AM_PROG_AR], [AM_PROG_AR])

dnl #
dnl # build flags
dnl #
AM_CPPFLAGS="-I\${top_srcdir}/include -I\${top_builddir}/include"
AM_CFLAGS="-Wall"
AM_LDFLAGS="-Wl,-z -Wl,relro"
AC_SUBST([AM_CPPFLAGS])
AC_SUBST([AM_CFLAGS])
AC_SUBST([AM_LDFLAGS])

dnl #
dnl # version information
dnl #
VERSION_MAJOR=$(echo ${VERSION} | cut -d'.' -f 1)
VERSION_MINOR=$(echo ${VERSION} | cut -d'.' -f 2)
VERSION_MICRO=$(echo ${VERSION} | cut -d'.' -f 3)
AC_SUBST([VERSION_MAJOR])
AC_SUBST([VERSION_MINOR])
AC_SUBST([VERSION_MICRO])

dnl #
dnl # Determine which C compiler to use
dnl #
AC_PROG_CXX
AC_PROG_CC
AM_PROG_CC_C_O

dnl #
dnl # Dependencies
dnl #
AC_CHECK_LIB([json-c], [json_object_object_get_ex], [],
	     [AC_MSG_ERROR([please install json-c])])

AC_CHECK_HEADERS([json-c/json.h], [],
		 [AC_MSG_ERROR([please install the json-c development package])])

LT_INIT

dnl #
dnl # Enable/Disable code coverage
dnl #
AX_CODE_COVERAGE

dnl #
dnl # makefiles
dnl #
AC_CONFIG_FILES([
	Makefile
	dist/Makefile
	doc/Makefile
	doc/examples/Makefile
	include/Makefile
	src/Makefile
	tests/Makefile
	tests/ftests/Makefile
	tests/gunit/Makefile
])
AC_CONFIG_FILES([dist/adaptived.spec:dist/adaptived.spec.in])

dnl #
dnl # generate the configure and makefiles
dnl #
AC_OUTPUT
