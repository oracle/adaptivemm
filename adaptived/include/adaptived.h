/*
 * Copyright (c) 2024, Oracle and/or its affiliates.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#ifndef __ADAPTIVED_H
#define __ADAPTIVED_H

#ifdef __cplusplus
extern "C" {
#endif

#include <json-c/json.h>
#include <stdbool.h>
#include <stdio.h>

/*
 * Opaque structure that contains cause information
 */
struct adaptived_cause;

/*
 * Opaque structure that contains effect information
 */
struct adaptived_effect;

/**
 * Opaque structure that contains the adaptived context
 */
struct adaptived_ctx;

/**
 * adaptived context attributes
 */
enum adaptived_attr {
	ADAPTIVED_ATTR_INTERVAL = 0, /* milliseconds */
	ADAPTIVED_ATTR_MAX_LOOPS,
	ADAPTIVED_ATTR_LOG_LEVEL,
	ADAPTIVED_ATTR_SKIP_SLEEP, /* for speeding up tests. do not use in production */
	ADAPTIVED_ATTR_RULE_CNT,
	ADAPTIVED_ATTR_DAEMON_MODE, /* run as daemon */
	ADAPTIVED_ATTR_DAEMON_NOCHDIR,
	ADAPTIVED_ATTR_DAEMON_NOCLOSE,

	ADAPTIVED_ATTR_CNT
};

enum adaptived_cgroup_value_type {
	ADAPTIVED_CGVAL_STR = 0,
	ADAPTIVED_CGVAL_LONG_LONG,
	ADAPTIVED_CGVAL_FLOAT,

	ADAPTIVED_CGVAL_CNT,
	/*
	 * When adaptived_cgroup_get_value() is invoked with ADAPTIVED_CGVAL_DETECT provided,
	 * adaptived will try to automatically detect the contents of the cgroup setting
	 * and return the appropriate value.
	 *
	 * This is an invalid setting for adaptived_cgroup_set_value()
	 */
	ADAPTIVED_CGVAL_DETECT
};

struct adaptived_cgroup_value {
	enum adaptived_cgroup_value_type type;

	union {
		char *str_value;
		long long ll_value;
		float float_value;
	} value;
};

struct adaptived_rule_stats {
	int cause_cnt;
	int effect_cnt;

	long long loops_run_cnt;
	long long trigger_cnt;
	long long snooze_cnt;
};

/**
 * Initialization routine for a cause
 * @param cse Cause structure for this cause
 * @param args_obj JSON object representation of the args parameter for this cause
 * @param interval How often in milliseconds the adaptived main loop will run
 */
typedef int (*adaptived_cause_init)(struct adaptived_cause * const cse, struct json_object *args_obj,
				 int interval);

/**
 * Main processing logic for a cause
 * @param cse Cause structure for this cause
 * @param time_since_last_run Delta time since this cause last ran
 *
 * Note that currently the adaptived_loop() passes in ctx->interval in the time_since_last_run
 * field.  In the future it could track the actual delta time between runs
 */
typedef int (*adaptived_cause_main)(struct adaptived_cause * const cse, int time_since_last_run);

/**
 * Exit routine for a cause
 * @param cse Cause structure for this cause
 *
 * This routine should clean up any memory allocated by the cause
 */
typedef void (*adaptived_cause_exit)(struct adaptived_cause * const cse);

/**
 * Cause functions structure
 */
struct adaptived_cause_functions {
	adaptived_cause_init init;
	adaptived_cause_main main;
	adaptived_cause_exit exit;
};

/**
 * Initialization routine for an effect
 * @param eff Effect structure for this effect
 * @param args_obj JSON object representation of the args parameter for this effect
 * @param cse Cause structure this effect is acting upon
 */
typedef int (*adaptived_effect_init)(struct adaptived_effect * const eff, struct json_object *args_obj,
			   const struct adaptived_cause * const cse);
/**
 * Main processing logic for an effect
 * @param eff Effect structure for this effect
 */
typedef int (*adaptived_effect_main)(struct adaptived_effect * const eff);
/**
 * Exit routine for an effect
 * @param eff Effect structure for this effect
 */
typedef void (*adaptived_effect_exit)(struct adaptived_effect * const eff);

/**
 * Effect functions structure
 */
struct adaptived_effect_functions {
	adaptived_effect_init init;
	adaptived_effect_main main;
	adaptived_effect_exit exit;
};

/**
 * Log an error message
 * @param fmt message format
 */
void adaptived_err(const char *fmt, ...);

/**
 * Log a warning message
 * @param fmt message format
 */
void adaptived_wrn(const char *fmt, ...);

/**
 * Log an informational message
 * @param fmt message format
 */
void adaptived_info(const char *fmt, ...);

/**
 * Log a debug message
 * @param fmt message format
 */
void adaptived_dbg(const char *fmt, ...);

/**
 * Initialize the adaptived context structure
 * @param config_file Path to the configuration file to be used
 *
 * @return pointer to valid adaptived context.  NULL on failure
 *
 * If config file is NULL, adaptived will use the default config file,
 * /etc/adaptived.json
 */
struct adaptived_ctx *adaptived_init(const char * const config_file);

/**
 * Destroy the adaptived context and release any resources
 * @param ctx adaptived context struct
 *
 * (*ctx) will be NULLed if it is a valid ctx pointer
 */
void adaptived_release(struct adaptived_ctx **ctx);

/**
 * Main adaptived processing loop
 * @param ctx the adaptived configuration context
 * @param parse Run the json parser against the context's config file
 */
int adaptived_loop(struct adaptived_ctx * const ctx, bool parse);

/**
 * Set an attribute in the adaptived context
 * @param ctx adaptived options struct
 * @param attr attribute name
 * @param value attribute value
 */
int adaptived_set_attr(struct adaptived_ctx * const ctx, enum adaptived_attr attr, uint32_t value);

/**
 * Get the value of an attribute in the adaptived context
 * @param ctx adaptived options struct
 * @param attr attribute name
 * @param value attribute value
 */
int adaptived_get_attr(struct adaptived_ctx * const ctx, enum adaptived_attr attr,
		    uint32_t * const value);

/*
 * Get the statistics for the specified rule
 * @param ctx adaptived options struct
 * @param name Rule name
 * @param stats Structure to store the stats info into
 */
int adaptived_get_rule_stats(struct adaptived_ctx * const ctx,
			  const char * const name, struct adaptived_rule_stats * const stats);

/**
 * Get the private data pointer in a cause structure
 * @param cse Cause pointer
 */
void *adaptived_cause_get_data(const struct adaptived_cause * const cse);

/**
 * Set the private data pointer in a cause structure
 * @param cse Cause pointer
 * @param data Private data structure pointer
 */
int adaptived_cause_set_data(struct adaptived_cause * const cse, void * const data);

/**
 * Get the private data pointer in an effect structure
 * @param eff Effect pointer
 */
void *adaptived_effect_get_data(const struct adaptived_effect * const eff);

/**
 * Set the private data pointer in an effect structure
 * @param eff Effect pointer
 * @param data Private data structure pointer
 */
int adaptived_effect_set_data(struct adaptived_effect * const eff, void * const data);

/**
 * Register a cause with adaptived
 * @param ctx adaptived context
 * @param name cause name
 * @param fns cause function handlers
 */
int adaptived_register_cause(struct adaptived_ctx * const ctx, const char * const name,
			  const struct adaptived_cause_functions * const fns);

/**
 * Register a effect with adaptived
 * @param ctx adaptived context
 * @param name effect name
 * @param fns effect function handlers
 */
int adaptived_register_effect(struct adaptived_ctx * const ctx, const char * const name,
			  const struct adaptived_effect_functions * const fns);

/**
 * Parse a string from the JSON configuration file
 * @param obj JSON object
 * @param key JSON object key
 * @param value pointer that on success will point at the requested string
 */
int adaptived_parse_string(struct json_object * const obj, const char * const key, const char **value);

/**
 * Parse an integer from the JSON configuration file
 * @param obj JSON object
 * @param key JSON object key
 * @param value destination to store the integer value
 */
int adaptived_parse_int(struct json_object * const obj, const char * const key, int * const value);

/**
 * Parse a float from the JSON configuration file
 * @param obj JSON object
 * @param key JSON object key
 * @param value destination to store the float value
 */
int adaptived_parse_float(struct json_object * const obj, const char * const key, float * const value);

/**
 * Parse a long long from the JSON configuration file
 * @param obj JSON object
 * @param key JSON object key
 * @param value destination to store the long long value
 */
int adaptived_parse_long_long(struct json_object * const obj, const char * const key,
			   long long * const value);

/**
 * Parse a cgroup value from the JSON configuration file
 * @param obj JSON object
 * @param key JSON object key
 * @param value destination to store the cgroup value
 *
 * @Note - it's the responsibility of the caller to free this structure by
 * calling adaptived_free_cgroup_value()
 */
int adaptived_parse_cgroup_value(struct json_object * const obj, const char * const key,
			      struct adaptived_cgroup_value * const value);

/**
 * Free a adaptived_cgroup_value structure
 * @param pointer to the struct to be freed
 */
void adaptived_free_cgroup_value(struct adaptived_cgroup_value * const val);

/**
 * Parse a boolean from the JSON configuration file
 * @param obj JSON object
 * @param key JSON object key
 * @param value destination to store the bool value
 */
int adaptived_parse_bool(struct json_object * const obj, const char * const key, bool * const value);

/**
 * Build a cause object in memory
 * @param name The name of the cause that will be inserted into the rule
 *
 * @return A pointer to a newly-allocated cause struct.  This cause is typically freed
 *         when the cause is added to a rule, but it can also be freed by calling
 *         adaptived_release_cause()
 *
 * Build a cause object in memory.  This function is currently only used if rules are being
 * added at runtime
 */
struct adaptived_cause *adaptived_build_cause(const char * const name);

/**
 * Free an in-memory cause
 * @param Pointer to the cause structure
 */
void adaptived_release_cause(struct adaptived_cause ** cse);

/**
 * Add a string key-value pair to an in-memory cause
 * @param cse Cause structure
 * @param key Corresponds to a JSON key in the args JSON field
 * @param value String value associated with the key
 */
int adaptived_cause_add_string_arg(struct adaptived_cause * const cse, const char * const key,
				const char * const value);

/**
 * Add an integer key-value pair to an in-memory cause
 * @param cse Cause structure
 * @param key Corresponds to a JSON key in the args JSON field
 * @param value Integer value associated with the key
 */
int adaptived_cause_add_int_arg(struct adaptived_cause * const cse, const char * const key, int value);

/**
 * Build a effect object in memory
 * @param name The name of the effect that will be inserted into the rule
 *
 * @return A pointer to a newly-allocated effect struct.  This effect is typically freed
 *         when the effect is added to a rule, but it can also be freed by calling
 *         adaptived_release_effect()
 *
 * Build a effect object in memory.  This function is currently only used if rules are being
 * added at runtime
 */
struct adaptived_effect *adaptived_build_effect(const char * const name);

/**
 * Free an in-memory effect
 * @param Pointer to the effect structure
 */
void adaptived_release_effect(struct adaptived_effect ** eff);

/**
 * Add a string key-value pair to an in-memory effect
 * @param eff Cause structure
 * @param key Corresponds to a JSON key in the args JSON field
 * @param value String value associated with the key
 */
int adaptived_effect_add_string_arg(struct adaptived_effect * const eff, const char * const key,
				const char * const value);

/**
 * Add an integer key-value pair to an in-memory effect
 * @param eff Cause structure
 * @param key Corresponds to a JSON key in the args JSON field
 * @param value Integer value associated with the key
 */
int adaptived_effect_add_int_arg(struct adaptived_effect * const eff, const char * const key, int value);

/**
 * Build a rule object in memory
 * @param name The name of the rule
 *
 * @return A pointer to the newly-allocated rule
 *
 * BUild a rule object in memory.  This function is currently only used if rules are being
 * added at runtime
 */
struct adaptived_rule *adaptived_build_rule(const char * const name);

/**
 * Free an in-memory rule
 * @param Pointer to the rule structure to be freed
 */
void adaptived_release_rule(struct adaptived_rule ** rule);

/**
 * Add a cause to an in-memory rule
 * @param rule Pointer to the rule object
 * @param cse Pointer to the cause object
 */
int adaptived_rule_add_cause(struct adaptived_rule * const rule, const struct adaptived_cause * const cse);


/**
 * Add an effect to an in-memory rule
 * @param rule Pointer to the rule object
 * @param eff Pointer to the effect object
 */
int adaptived_rule_add_effect(struct adaptived_rule * const rule, const struct adaptived_effect * const eff);

/**
 * Given an in-memory rule, load it into the adaptived loop for processing
 * @param ctx adaptived context
 * @param rule In-memory rule to be loaded
 */
int adaptived_load_rule(struct adaptived_ctx * const ctx, struct adaptived_rule * const rule);

/**
 * Unload a rule from the adaptived processing loop
 * @param ctx adaptived context
 * @param name Rule name to be unloaded
 */
int adaptived_unload_rule(struct adaptived_ctx * const ctx, const char * const name);

#ifdef __cplusplus
} /* extern "C" */
#endif

#endif /* __ADAPTIVED_H */
