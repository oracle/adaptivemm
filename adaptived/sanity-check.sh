#!/bin/bash
# Copyright (c) 2024, Oracle and/or its affiliates.
# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
#
# This code is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 only, as
# published by the Free Software Foundation.
#
# This code is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# version 2 for more details (a copy is included in the LICENSE file that
# accompanied this code).
#
# You should have received a copy of the GNU General Public License version
# 2 along with this work; if not, write to the Free Software Foundation,
# Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
# or visit www.oracle.com if you need additional information or have any
# questions.

# clean up from a previous run
make clean
find . -name "*.gcno" -exec sudo rm '{}' \;
find . -name "*.gcda" -exec sudo rm '{}' \;

./autogen.sh
RET=$?

if [[ $RET -ne 0 ]]; then
	echo "./autogen failed: " $RET
	exit $RET
fi

CFLAGS="$CFLAGS -g -O0 -Werror" ./configure
RET=$?

if [[ $RET -ne 0 ]]; then
	echo "./configure failed: " $RET
	exit $RET
fi

make distcheck
RET=$?

if [[ $RET -ne 0 ]]; then
	echo "make distcheck failed: " $RET
	exit $RET
fi

CFLAGS="$CFLAGS -g -O0 -Werror" ./configure --enable-code-coverage
RET=$?

if [[ $RET -ne 0 ]]; then
	echo "./configure w/ code coverage failed: " $RET
	exit $RET
fi

# I'm not fond of running all of the tests as root, but the sd-bus tests
# need administrative privileges.  Merging the non-sudo and sudo code
# coverage files is difficult.  Let's avoid that.
sudo make check
RET=$?

if [[ $RET -ne 0 ]]; then
	echo "make check failed: " $RET
	exit $RET
fi

gcovr -s --filter src/
exit 0
