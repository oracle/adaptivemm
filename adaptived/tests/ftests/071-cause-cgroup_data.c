/*
 * Copyright (c) 2024-2025, Oracle and/or its affiliates.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
/**
 * Test to exercise the cgroup_data cause
 *
 * Note that this test creates a fake cgroup hierarchy directly in the
 * tests/ftests directory and operates on it.
 *
 */

#include <sys/stat.h>
#include <sys/wait.h>
#include <stdbool.h>
#include <unistd.h>
#include <syslog.h>
#include <string.h>
#include <signal.h>
#include <errno.h>

#include <adaptived.h>

#include "ftests.h"

#define EXPECTED_RET -ETIME

static const char * const cgroup_dirs[] = {
	"./test071cgroup",
	"./test071cgroup/child1",
	"./test071cgroup/child1/grandchild11/",
	"./test071cgroup/child2",
	"./test071cgroup/child3"
};
static const int cgroup_dirs_cnt = ARRAY_SIZE(cgroup_dirs);

static const char * const cgroup_files[] = {
	"./test071cgroup/memory.current",
	"./test071cgroup/cpu.uclamp.min",
	"./test071cgroup/cpuset.cpus.effective",
	"./test071cgroup/cpu.max",

	"./test071cgroup/child1/memory.current",
	"./test071cgroup/child1/cpu.uclamp.min",
	"./test071cgroup/child1/cpuset.cpus.effective",
	"./test071cgroup/child1/cpu.max",

	"./test071cgroup/child1/grandchild11/memory.current",
	"./test071cgroup/child1/grandchild11/cpu.uclamp.min",
	"./test071cgroup/child1/grandchild11/cpuset.cpus.effective",
	"./test071cgroup/child1/grandchild11/cpu.max",

	"./test071cgroup/child2/memory.current",
	"./test071cgroup/child2/cpu.uclamp.min",
	"./test071cgroup/child2/cpuset.cpus.effective",
	"./test071cgroup/child2/cpu.max",

	"./test071cgroup/child3/memory.current",
	"./test071cgroup/child3/cpu.uclamp.min",
	"./test071cgroup/child3/cpuset.cpus.effective",
	"./test071cgroup/child3/cpu.max",
};
static const int cgroup_files_cnt = ARRAY_SIZE(cgroup_files);

/*
 * Write the file contents as strings since that's how adaptived will read out
 * the data.  adaptived will then try to determine the correct data type.
 */
static const char * const file_contents[] = {
	"4096000",
	"1.234",
	"1-2,4-7,10,12",
	"max 100000",

	"8192000",
	"3.456",
	"1-20",
	"10000 100000",

	"1024000",
	"0.000",
	"1,3,5,7,9,11",
	"250000 1000000",

	"0",
	"9.876",
	"5-8",
	"max 1000000",

	"12345678",
	"4.680",
	"8,10",
	"500000 1000000",
};
static_assert(ARRAY_SIZE(file_contents) == ARRAY_SIZE(cgroup_files),
	      "file_contents should be the same size as cgroup_files_cnt");


static void write_cgroup_files(void)
{
	int i;

	for (i = 0; i < cgroup_files_cnt; i++)
		write_file(cgroup_files[i], file_contents[i]);
}

int main(int argc, char *argv[])
{
	char config_path[FILENAME_MAX], expected_file[FILENAME_MAX];
	struct adaptived_ctx *ctx = NULL;
	int ret;

	snprintf(config_path, FILENAME_MAX - 1, "%s/071-cause-cgroup_data.json",
		 argv[1]);
	config_path[FILENAME_MAX - 1] = '\0';

	ctx = adaptived_init(config_path);
	if (!ctx)
		return AUTOMAKE_HARD_ERROR;

	ret = create_dirs(cgroup_dirs, cgroup_dirs_cnt);
	if (ret)
		goto err;

	write_cgroup_files();

	ret = adaptived_set_attr(ctx, ADAPTIVED_ATTR_MAX_LOOPS, 1);
	if (ret)
		goto err;
	ret = adaptived_set_attr(ctx, ADAPTIVED_ATTR_INTERVAL, 1234);
	if (ret)
		goto err;
	ret = adaptived_set_attr(ctx, ADAPTIVED_ATTR_SKIP_SLEEP, 1);
	if (ret)
		goto err;
	ret = adaptived_set_attr(ctx, ADAPTIVED_ATTR_LOG_LEVEL, LOG_EMERG);
	if (ret)
		goto err;

	ret = adaptived_loop(ctx, true);
	if (ret != EXPECTED_RET) {
		adaptived_err("Test 071 returned: %d, expected: %d\n", ret, EXPECTED_RET);
		goto err;
	}

	adaptived_release(&ctx);
	ctx = NULL;

	snprintf(expected_file, FILENAME_MAX - 1, "%s/071-cause-cgroup_data.expected", argv[1]);
	expected_file[FILENAME_MAX - 1] = '\0';

	ret = compare_files_unsorted("071-cause-cgroup_data.out", expected_file);
	if (ret)
		goto err;

	delete_file("071-cause-cgroup_data.out");
	delete_files(cgroup_files, cgroup_files_cnt);
	delete_dirs(cgroup_dirs, cgroup_dirs_cnt);

	return AUTOMAKE_PASSED;

err:
	delete_file("071-cause-cgroup_data.out");
	delete_files(cgroup_files, cgroup_files_cnt);
	delete_dirs(cgroup_dirs, cgroup_dirs_cnt);
	if (ctx)
		adaptived_release(&ctx);

	return AUTOMAKE_HARD_ERROR;
}
