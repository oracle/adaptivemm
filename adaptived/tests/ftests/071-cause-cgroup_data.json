{
	"rules": [
		{
			"name": "Gather cgroup data and print it out",
			"causes": [
				{
					"name": "cgroup_data",
					"args": {
						"cgroup": "./test071cgroup///",
						"settings": [
							{
								"setting": "memory.current"
							},
							{
								"setting": "cpu.uclamp.min"
							},
							{
								"setting": "cpuset.cpus.effective"
							}
						]
					}
				}
			],
			"effects": [
				{
					"name": "print",
					"args": {
						"file": "071-cause-cgroup_data.out",
						"shared_data": true
					}
				}
			]
		}
	]
}
