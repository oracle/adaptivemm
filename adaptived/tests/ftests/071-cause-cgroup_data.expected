Print effect triggered by:
	cgroup_data
Cause "cgroup_data" shared
	cgroup "child1"
	setting "memory.current"
	long long value "8192000"
Cause "cgroup_data" shared
	cgroup "child1"
	setting "cpu.uclamp.min"
	float value "3.456000"
Cause "cgroup_data" shared
	cgroup "child1"
	setting "cpuset.cpus.effective"
	string value "1-20"
Cause "cgroup_data" shared
	cgroup "child2"
	setting "memory.current"
	long long value "0"
Cause "cgroup_data" shared
	cgroup "child2"
	setting "cpu.uclamp.min"
	float value "9.876000"
Cause "cgroup_data" shared
	cgroup "child2"
	setting "cpuset.cpus.effective"
	string value "5-8"
Cause "cgroup_data" shared
	cgroup "child3"
	setting "memory.current"
	long long value "12345678"
Cause "cgroup_data" shared
	cgroup "child3"
	setting "cpu.uclamp.min"
	float value "4.680000"
Cause "cgroup_data" shared
	cgroup "child3"
	setting "cpuset.cpus.effective"
	string value "8,10"
